LightGBM: Upgrade to Fast Predict API and fix double free bug

AlbertoEAF opened the pull request:
* Use the newer prediction *Fast() methods and FastConfig
* Reduce temporary object creation (less GC in prediction for better latencies)
* Number of features in prediction classes is now computed from the model (cleaner)
* Visibility refactor for methods in the SWIG helper classes
* Add SWIGResourcesTest and SWIGResourcesTest classes to test resource initialization/release
* Found and fixed 2 double free bugs in SWIGResources
* Instead of hardcoded in LightGBMSWIG, LightGBM prediction parameters are passed to the constructor

shenggwang commented on openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMSWIG.java:
```diff
@@ -88,16 +76,17 @@
      */
     LightGBMSWIG(final String modelPath, final DatasetSchema schema) throws ModelLoadingException {
+
+        final String LightGBMParameters = "num_threads=1";
```
If one day I want to add more parameters, like [device_type]( https://lightgbm.readthedocs.io/en/latest/Parameters.html#device_type), how could I do that?
is it "num_threads=1, device_type=cpu"?
I think we should change this in order to be flexible for further changes

AlbertoEAF replied:
Yes it is :) A single string. I don't think we have parameters to use there.
What do you suggest changing? Or given that it's just a string it's good?

shenggwang replied:
I think we can define it within the class, but out of the constructor as a private final static String and maybe explain it on javadoc that we can add more parameters by adding comma.
WDYT?

AlbertoEAF replied:
Hmm I like the idea but if you really want to modify this, it would likely be passed as a parameter to the constructor so each model could use its custom parameters.

Making it static would go against it.

shenggwang replied:
I think it would be better then.
Then you will have to create the variable on the constructor of LightGBMBinaryClassificationModel, but this time it can be a static variable right?

shenggwang commented on openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMSWIG.java:
```diff
@@ -132,71 +121,38 @@ private void initSWIGInstanceArray(final Instance instance) {
             if (returnCodeLGBM == -1)
                 throw new LightGBMException();
 
+            // Valid for binary predictions:
```
The comment isn't clear
is it predicting a valid binary score?

AlbertoEAF replied:
True, will remove :p

shenggwang commented on openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/SWIGResources.java:
```diff
@@ -76,19 +82,41 @@
      */
-    SWIGResources(final String modelPath, final int numFeaturesSchema) throws ModelLoadingException, LightGBMException {
+    SWIGResources(final String modelPath,
+                  final String LightGBMParameters) throws ModelLoadingException, LightGBMException {
```
Javadoc is changed to remove the previous parameter, but you should document the new one

shenggwang commented on openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMSWIG.java:
```diff
@@ -88,16 +76,17 @@
      * @throws ModelLoadingException in case any LightGBM error occurs.
      */
     LightGBMSWIG(final String modelPath, final DatasetSchema schema) throws ModelLoadingException {
```
I saw that you had provided all method and variable `private` and `public` attribute, why don't you give public to class and class constructor?

AlbertoEAF replied:
We don't want to make this low-level helper classes visible outside the package.

I can add public to the constructor though :+1:

shenggwang commented on openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/SWIGResources.java:
```diff
@@ -76,19 +82,41 @@
      * @throws LightGBMException     in case there's an error in the C++ core library.
      */
-    SWIGResources(final String modelPath, final int numFeaturesSchema) throws ModelLoadingException, LightGBMException {
+    SWIGResources(final String modelPath,
```
Same public attribute question as above

shenggwang reviewed (commented):
I don't see UnitTest to validate these changes. Is it testable?

AlbertoEAF commented:
Yes you're right @shenggwang , I added a new SWIGResourcesTest class for this was never tested and it's what's preventing memory leaks etc.
Although we are not testing directly that we're calling the appropriate SWIG delete_*() methods for each field, we ensure all fields are initialized and set back to null at the end (and in the code the set back to null only happens after calling delete_*()).

With this 2 new bugs related to "double-free" when calling close() in C++ were found and fixed.

Regarding the other changes to the predict API, the scoring tests already cover that our scores are correct against the reference scores ;)

AlbertoEAF commented:
@shenggwang @jlazevedo should I give the same public/private treatment to SWIGTrainResources + test class?

The class is defined as public when it should have the default visibility. Coincidently I don't see the double-free mistake as with SWIGResources, although it has no tests either.

shenggwang commented on openml-lightgbm/lightgbm-provider/src/test/java/com/feedzai/openml/provider/lightgbm/SWIGResourcesTest.java:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * Standard model path.
+     */
+    static Path modelPath;
```
Can this be private?

shenggwang commented on openml-lightgbm/lightgbm-provider/src/test/java/com/feedzai/openml/provider/lightgbm/SWIGResourcesTest.java:
```diff
@@ -0,0 +1,122 @@
+     * @throws URISyntaxException Couldn't extract file resource.
+     */
+    @BeforeClass
+    static public void setupFixture() throws ModelLoadingException, URISyntaxException {
```
change the order to public static

shenggwang commented on openml-lightgbm/lightgbm-provider/src/test/java/com/feedzai/openml/provider/lightgbm/SWIGResourcesTest.java:
```diff
@@ -0,0 +1,122 @@
+     * SWIGResources instance common to all tests. Do not modify it!
+     * Used to speed up tests by avoiding re-reading the model file.
+     */
+    static SWIGResources defaultSwig;
```
same as above

shenggwang commented:
> @shenggwang @jlazevedo should I give the same public/private treatment to SWIGTrainResources + test class?
> 
> The class is defined as public when it should have the default visibility. Coincidently I don't see the double-free mistake as with SWIGResources, although it has no tests either.

Are you saying that SWIGResources shouldn't be public?
In that case, can it be protected? [pacakge-private vs protected](https://stackoverflow.com/a/215579)

AlbertoEAF commented:
> > @shenggwang @jlazevedo should I give the same public/private treatment to SWIGTrainResources + test class?
> > The class is defined as public when it should have the default visibility. Coincidently I don't see the double-free mistake as with SWIGResources, although it has no tests either.
> 
> Are you saying that SWIGResources shouldn't be public?
> In that case, can it be protected? [pacakge-private vs protected](https://stackoverflow.com/a/215579)

Following the principle of least-visibility, it would make more sense here to be package-private. This is still low-level code not meant to be modified/extended/used elsewhere.

Should we add tests?

shenggwang commented:
> > > @shenggwang @jlazevedo should I give the same public/private treatment to SWIGTrainResources + test class?
> > > The class is defined as public when it should have the default visibility. Coincidently I don't see the double-free mistake as with SWIGResources, although it has no tests either.
> > 
> > 
> > Are you saying that SWIGResources shouldn't be public?
> > In that case, can it be protected? [pacakge-private vs protected](https://stackoverflow.com/a/215579)
> 
> Following the principle of least-visibility, it would make more sense here to be package-private. This is still low-level code not meant to be modified/extended/used elsewhere.
> 
> Should we add tests?

If it is not intended to be accessed besides this package. So, left it as default

AlbertoEAF commented:
Travis CI build failing since LLVM's artifact with clang for ubuntu 14.04 is down. Let's hope it's just an outage.

jlazevedo reviewed (changes requested):
Good effort doing some refactoring. 
Do not loose too much time with `public vs package-private vs protected` discussion. The project has a single package and some tight coupling, we can think about that further down the road.

shenggwang commented on openml-lightgbm/lightgbm-provider/src/test/java/com/feedzai/openml/provider/lightgbm/SWIGResourcesTest.java:
```diff
@@ -0,0 +1,122 @@
+ * Tests for the SWIGResources class to ensure all resources are properly/safely initialized and released.
+ *
+ * @author Alberto Ferreira (alberto.ferreira@feedzai.com)
+ * @since 1.0.18
```
AFAIK, we don't have a replacement for this in the open-source projects. I think this may be a future work

AlbertoEAF replied:
The test class is new.

shenggwang replied:
Note, by "this" I meant `@since` tag

shenggwang commented on openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/SWIGTrainResources.java:
```diff
@@ -34,7 +34,7 @@
  * @since 1.0.10
  */
-public class SWIGTrainResources implements AutoCloseable {
+class SWIGTrainResources implements AutoCloseable {
```
The last comment, you made class package-private you should have also made constructor package-private.
